### PR TITLE
Autorun

### PR DESCRIPTION
**Pull request by tyoungsc**
# Adding a New Sample(s)
## Description
A new FPGA tutorial to use the generic autorun header. This utility is mostly to help users coming from OpenCL who used autorun kernels there.

### CODE REVIEW

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/README.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
+# Autorun Kernels
+This FPGA tutorial demonstrates how to create autorun kernels. These kernels are submitted before the main function begins and typically loop forever.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/README.md]: ...create the equivalent of OpenCL 'autorun' kernels in oneAPI.  An autorun kernel is one that is submitted before the main() function begins.  Typically autorun kernels will never finish.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/README.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+| Software                          | Intel® oneAPI DPC++ Compiler <br> Intel® FPGA Add-On for oneAPI Base Toolkit
+| What you will learn               | How and when to use autorun kernels
+| Time to complete                  | 15 minutes
+
+## Purpose
+The purpose of this tutorial is to demonstrate how to create autorun kernels in DPCPP. Autorun kernels are submitted to the SYCL queue automatically before `main()` begins and are meant to have no direct interaction with the host. This means that they should not directly access global memory (e.g., via USM pointers or SYCL acccessors) and the host should not wait for them to finish.
+
+Typically, these kernels are meant to run forever, and data is streamed to and from them using producer and consumer kernels and SYCL pipes. This technique is illustrated in the figures below where the middle light-blue kernels (`ARKernel` and `ARForeverKernel`) are autorun kernels, and the dark-blue kernels on the left and right are regular SYCL kernels. The images below correspond to the `Autorun` and `AutorunForever` kernels used in this tutorial. See the comments and code in `autorun.cpp` for more details on the differences.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/README.md]: I think 'producer and consumer kerenls' should be deleted here?  Just 'streamed to and from them using SYCL pipes'?

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.cpp [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+// we make this a global variable so it can be used by the autorun kernels
+#if defined(FPGA_EMULATOR)
+  ext::intel::fpga_emulator_selector ds;
+#else
+  ext::intel::fpga_selector ds;
+#endif
+
+// declare the kernel names globally to reduce name mangling
+class ARProducerId;
+class ARKernelID;
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.cpp:26]: You have a mix of Id and ID, might as well pick one for consistency

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.cpp [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+struct MyAutorun {
+  void operator()() const {
+    // notice that in this version, we explicitly add the while(1)-loop
+    while (1) {
+      auto d = ARProducePipe::read();
+      ARConsumePipe::write(d);
+    }
+  }
+};
+fpga_tools::Autorun<ARKernelID> ar_kernel{ds, MyAutorun{}};
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.cpp:58]: I'd put a comment on this line (and line 69 below), something like 'Declaring a global instance of this class causes the constructor to be called before main() runs, and the constructor launches the kernel'

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.cpp [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+using ARProducePipe = ext::intel::pipe<ARProducePipeID, int>;
+using ARConsumePipe = ext::intel::pipe<ARConsumePipeID, int>;
+using ARForeverProducePipe = ext::intel::pipe<ARForeverProducePipeID, int>;   
+using ARForeverConsumePipe = ext::intel::pipe<ARForeverConsumePipeID, int>;
+
+////////////////////////////////////////////////////////////////////////////////
+// Autorun user kernel and global variable
+struct MyAutorun {
+  void operator()() const {
+    // notice that in this version, we explicitly add the while(1)-loop
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.cpp:48]: This implies that an autorun kernel must run forever, either with an explicit or an implicit while(1).  Are there use cases for Autorun where the kernel might do a limited amout of stuff then finish?  I can't think of anything useful off the top of my head . . .

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+#include <CL/sycl.hpp>
+#include <type_traits>
+
+/*
+This header defines the Autorun kernel utility. This utility is used to
+launch kernels that are submitted before main begins. It is typically used
+to launch kernels that run forever.
+
+Two classes are defined in this header file: Autorun and AutorunForever.
+Autorun creates an autorun kernel that is NOT implicitly wrapped in an infinite
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp:13]: I would rephrase this.  AutorunForever creates a kernel and wraps it in a while(1) loop.  Autorun creates the kernel and does not wrap it in a while(1) loop.  Usually when using Autorun the user will have the while(1) loop explicitly in their code.  Something like that.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+                   "KernelFunctor must be callable with no arguments");
+
+      // create the device queue
+      sycl::queue q{device_selector};
+
+      // run the single task kernel
+      q.single_task<KernelID>(kernel);
+    }
+
+    // Constructor without a kernel name
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp]: There's a lot of duplicated code here, is there a clever way we could combine these constructors?

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+      // static asserts to ensure KernelFunctor is callable
+      static_assert(std::is_invocable_r_v<void, KernelFunctor>,
+                   "KernelFunctor must be callable with no arguments");
+      
+      // create the device queue
+      sycl::queue q{device_selector};
+
+      // run the single task
+      q.single_task<KernelID>([=] {
+        while (1) {
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp]: Not sure this is worth doing, just an idea.  You could combine Autorun and AutorunForever by adding a template parameter to Autorun <bool k_run_forever>, then rather than while(1) you would have do { } while (k_run_forever).  This would cut out a lot of duplicated code.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.cpp (hunk 5 above)
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.cpp:48]: Joe wanted these because he likes the explicit while-loop. And it at least gives the user an option to not have the while(1)

**Comment by tyoungsc** [comment]
Addressed your comments @mdbtucker

**Review by mdbtucker — APPROVED** [review]
Looks good, please create a PR to the main repo.

**Comment by mdbtucker** [comment]
Closing this PR so you don't accidentally merge it to your fork's development branch.
